falcon/api.py

```diff
@@ -728,14 +728,22 @@ def _compile_uri_template(template):
         if template != '/' and template.endswith('/'):
             template = template[:-1]
 
-        expression_pattern = r'{([a-zA-Z][a-zA-Z_]*)}'
+        PCT_ENCODED = '%[0-9A-Fa-f]{2}'
+        UNRESERVED = '[^/]'
+        RESERVED = r'[:/\?#\[\]@\!\$&\'\(\)\*\+,;=]'
+        BASIC_VARIABLE = '{([a-zA-Z][a-zA-Z_]+)}'
+        RESERVED_VARIABLE = r'{\+([a-zA-Z][a-zA-Z_]+)}'
```

Comment:
Please change the ```[a-zA-Z_]``` on L734-735 to ```\w``` as that allows 0-9 in the remainder of the variable names.
I did that in https://github.com/falconry/falcon/pull/454 for ```expression_pattern``` which you are replacing.